datadog_checks_base/datadog_checks/base/utils/agent/memory.py

```diff
@@ -0,0 +1,275 @@
+
+    with open(path, 'w', encoding='utf-8') as f:
+        f.write('Top {} lines\n'.format(limit))
+
+        for index, stat in enumerate(top_stats[:limit], 1):
+            frame = stat.traceback[MOST_RECENT_FRAME]
+
+            amount, unit = unit_formatter(stat.size)
+            f.write(
+                '\n#{}: {}:{}: {} {}\n'.format(index, parse_package_path(frame.filename), frame.lineno, amount, unit)
+            )
+
+            line = linecache.getline(frame.filename, frame.lineno).strip()
+            if line:
+                f.write(' {}  {}\n'.format(' ' * len(str(index)), line))
+
+        f.write('\n')
+
+        other = top_stats[limit:]
+        if other:
+            size = sum(stat.size for stat in other)
+            amount, unit = unit_formatter(size)
+            f.write('{} other: {} {}\n'.format(len(other), amount, unit))
+
+        total = sum(stat.size for stat in top_stats)
+        amount, unit = unit_formatter(total)
+        f.write('Total allocated size: {} {}\n'.format(amount, unit))
+
+
+def write_pretty_diff(
+    path, current_snapshot, previous_snapshot, unit_formatter, key_type, limit, cumulative, diff_order
+):
+    top_stats = current_snapshot.compare_to(previous_snapshot, key_type=key_type, cumulative=cumulative)
+
+    if diff_order == 'positive':
+        # Keep the default behavior, but add precedence for memory increases
+        top_stats = sorted(top_stats, key=lambda stat: (int(stat.size_diff > 0),) + stat._sort_key())[::-1]
+
+    with open(path, 'w', encoding='utf-8') as f:
+        f.write('Top {} line diffs\n'.format(limit))
+
+        index = 0
+        for stat in top_stats[:limit]:
+            # Disregard lines that have no diff as the top lines are already shown by the snapshot files
+            if not stat.size_diff:
+                continue
+
+            index += 1
+            frame = stat.traceback[MOST_RECENT_FRAME]
+
+            amount, unit = unit_formatter(abs(stat.size_diff))
+            f.write(
+                '\n#{}: {}:{}: {}{} {}\n'.format(
+                    index, parse_package_path(frame.filename), frame.lineno, get_sign(stat.size_diff), amount, unit
+                )
+            )
+
+            line = linecache.getline(frame.filename, frame.lineno).strip()
+            if line:
+                f.write(' {}  {}\n'.format(' ' * len(str(index)), line))
+
+        f.write('\n')
+
+        other = top_stats[limit:]
+        if other:
+            size = sum(stat.size_diff for stat in other)
+            amount, unit = unit_formatter(abs(size))
+            f.write('{} other: {}{} {}\n'.format(len(other), get_sign(size), amount, unit))
+
+        total = sum(stat.size_diff for stat in top_stats)
+        amount, unit = unit_formatter(abs(total))
+        f.write('Total difference: {}{} {}\n'.format(get_sign(total), amount, unit))
+
+
+def profile_memory(f, config, namespaces=None, args=(), kwargs=None):
```

Comment:
How is `profile_memory` used ? Just dropping a line like this in the code ? 

```python
profile_memory(fn_to_profile, instance_config, [arg1, arg2], {"a":"b"})
```

Reply:
Yup! https://github.com/DataDog/integrations-core/pull/4166/files#diff-874d5159f166811f6c78ff097251312cR507